ncm-opennebula/src/main/perl/OpenNebula/Network.pm

```diff
@@ -0,0 +1,207 @@
+}
+
+=item get_vnetars
+
+Gets the network C<ARs> (address range) from tt file
+and gathers C<VNet> names and IPs/MAC addresses.
+
+=cut
+
+sub get_vnetars
+{
+    my ($self, $config) = @_;
+    my $all_ars = $self->process_template_aii($config, "network_ar");
+    my %res;
+
+    my @tmp = split(qr{^NETWORK\s+=\s+(?:"|')(\S+)(?:"|')\s*$}m, $all_ars);
+
+    while (my ($ar, $network) = splice(@tmp, 0, 2)) {
+        if ($network && $ar) {
+            $self->verbose("Detected network AR: $ar within network $network");
+            $res{$network}{ar} = $ar;
+            $res{$network}{network} = $network;
+            $self->debug(3, "This is the network AR template for $network: $ar");
+        } else {
+            # No ars found for this VM
+            $self->error("No network ARs and/or network info $ar.");
+        };
+    }
+    return %res;
+}
+
+=item remove_and_create_vn_ars
+
+Removes/creates C<ARs> (address range).
+
+=cut
+
+sub remove_and_create_vn_ars
+{
+    my ($self, $one, $arsref, $remove) = @_;
+
+    foreach my $vnet (sort keys %{$arsref}) {
+        my $ardata = $arsref->{$vnet};
+        $self->info ("Testing ONE vnet network AR: $vnet");
+
+        my %ar_opts = ('template' => $ardata->{ar});
+        my @exisvnet = $one->get_vnets(qr{^$vnet$});
+        foreach my $vnar (@exisvnet) {
+            my $arinfo = $vnar->get_ar(%ar_opts);
+            if ($remove) {
+                $self->remove_duplicate_ars($one, $vnet, $vnar, $arinfo);
+                $self->remove_vn_ars($one, $arinfo, $vnet, $ardata, $vnar);
+            } else {
+                # At this moment it is not possible to update ONE ARs (only leases number)
+                # you should remove and recreate the AR/VM if something changes
+                if ($arinfo) {
+                    $self->remove_duplicate_ars($one, $vnet, $vnar, $arinfo);
+                    $self->verbose("vnet: $vnet already contains AR id: ", $arinfo->{AR_ID}->[0]);
+                } else {
+                    $self->create_vn_ars($one, $vnet, $ardata, $vnar);
+                };
+            }
+       }
+    }
+}
+
+=item detect_duplicate_ars
+
+Detects duplicate C<VirtualNetwork> C<ARs> with
+same IPs or MACs.
+Removes duplicated C<ARs> (if C<QUATTOR> flag is set to true).
+
+=cut
+
+sub remove_duplicate_ars
+{
+    my ($self, $one, $vnet, $vnar, $arinfo) = @_;
+
+    # Try to find different AR ids with the same configuration
+    if (defined($arinfo)) {
+        my %arpool = $vnar->get_ar_pool();
+        my $mac = $arinfo->{MAC}->[0];
+        my $ip = $arinfo->{IP}->[0];
+        my $arid = $arinfo->{AR_ID}->[0];
+        foreach my $id (sort keys %arpool) {
+            if (($arpool{$id}->{MAC}->[0] eq $mac) && ($id != $arid)) {
+                $self->warn("Current AR $arid and $vnet AR $id have the same MAC $mac");
+                $self->remove_vn_ars($one, $arpool{$id}, $vnet, $arinfo, $vnar);
+            } elsif (($arpool{$id}->{IP}->[0] eq $ip) && ($id != $arid)) {
```

Comment:
I think so, we only have to avoid 3 cases: 
a) everything is the same but with different `id`  (that's not good but could happen at this point)
b) same MAC with different `id` or 
c) same IP with different `id`  .